llvm/lib/Analysis/ValueTracking.cpp

```diff
@@ -2420,9 +2420,20 @@ static bool isNonZeroRecurrence(const PHINode *PN) {
   }
 }
 
+static bool matchOpWithOpEqZero(Value *Op0, Value *Op1) {
+  ICmpInst::Predicate Pred;
+  return (match(Op0, m_ZExtOrSExt(m_ICmp(Pred, m_Specific(Op1), m_Zero()))) ||
+          match(Op1, m_ZExtOrSExt(m_ICmp(Pred, m_Specific(Op0), m_Zero())))) &&
+         Pred == ICmpInst::ICMP_EQ;
+}
+
 static bool isNonZeroAdd(const APInt &DemandedElts, unsigned Depth,
                          const SimplifyQuery &Q, unsigned BitWidth, Value *X,
                          Value *Y, bool NSW, bool NUW) {
+  // (X + (X != 0)) is non zero
```

Comment:
Yeah, it is a common pattern. But I would like to canonicalize it into `umax(X, 1)` (or vice versa).

Reply:
Think umax would make sense, although imo this patch still makes sense b.c this will be a single-user only transform

Reply:
Would umax canonicalization cover the original SPEC case?

Reply:
yes

Reply:
I think there are two transform here.
1) fold `(X == C) Op X` -> `X == C ? (X Op 0) : (C Op 1)`. assuming that `X Op 0` simplifies.
2) recognize something like `X == 0 ? X : 1` as `umax(X, 1)`.
I'm happy to write patches for both.

Reply:
We already handle the `X == 0 ? 1 : X` -> `umax(X, 1)`. Ill post patch for the select case soon. That being said, still prefer to get this in as I don't think we should rely on single-use constrained canonicalization in `ValueTracking`.

Reply:
Made PR for the folding #89020
Would still like to get this in for the reason of the fold requiring single-use.